### ISSUE TITLE
`nextion` service calls won't work in container deployments of Home Assistant

### ISSUE THREAD

**Issue by cmuellner** [Blackymas/NSPanel_HA_Blueprint#1195]
If I understood this blueprint correctly, then the blueprint depends on the `nextion` service (e.g. `service: '{{ nextion.command.tft_upload }}'`). However, this service is not part of Home Assistant, but is an add-on of the HASS distribution.

So people who use Home Assistant in a deployment, which does not support add-ons (see [HA installation overview](https://www.home-assistant.io/installation#compare-installation-methods)) can install the blueprint, but will end up with a non-functioning blueprint, that does not even state why it is not working.

Note, that in such an environment, where Home Assistant is running in a container, ESPHome is likely running in a different container. There exists a Home Assistant Integration for ESPHome devices, but that does only access the devices, not the server.

At minimum it would be great to have HASS documented as dependency (Home Assistant is not enough).

A better solution would be to be able to generate the TFT files in another way and use the `esphome.{{ nspanel_name }}_upload_tft'` command to upload.

The best solution, that I can think of is to have code and documentation to make this blueprint work in a containerized environment.

Thoughts and suggestions are welcome.

**Comment by edwardtfn** [comment]
> A better solution would be to be able to generate the TFT files in another way and use the esphome.{{ nspanel_name }}_upload_tft' command to upload.

This is exactly how it works.

**Comment by cmuellner** [comment]
> > If I understood this blueprint correctly, then the blueprint depends on the nextion service (e.g. service: '{{ nextion.command.tft_upload }}'). However, this service is not part of Home Assistant, but is an add-on of the HASS distribution.
> 
> That is not correct. There's no dependencies to any add-on on Home Assistant. It uses ESPHome just to compile the code, but even if you don't have ESPHome as an add-on to HASS you can still compile it externally and use the panel with Home Assistant.

Are you sure about that, because all documentation and videos say that ESPhome needs to be installed as add-on (and not as integration).

> Upload TFT is one method of the Nextion component included in ESPHome and the final code is totally embedded in the firmware loaded to the panel. It's recommend to use a local http server for transferring the TFT file, but again, if you don't have that, it would be possible to transfer it directly from GitHub (just be prepared for a few retries in that case).

I managed to compile the ESPHome firmware and upload the firmware (UART and WiFi).

Also, the container deployment of HA also provides the `/config/www`/`local` HTTP file server.
So updating the TFT file over WiFi works fine.

What does not work is the generation of the TFT file as configured in the blueprint.

I believe the reason for that is, that service calls to the `nextion` service don't work.
There are 92 calls to the `nextion` service in the blueprint, so this has some impact (search for `service: '{{ nextion.` in the [blueprint](https://github.com/Blackymas/NSPanel_HA_Blueprint/blob/main/nspanel_blueprint.yaml)).
Also, if I go to "Developer Tools" -> "Services" there is nothing found if I type `nextion` in the search box.
However, if I type "esphome" it shows the services that the panel is exposing.

So I guess my question is: how can I get this `nextion` service that is used by the blueprint?

**Comment by edwardtfn** [comment]
> Are you sure about that, because all documentation and videos say that ESPhome needs to be installed as add-on (and not as integration).

I'm sure about that. The videos are showing the easier way for most users, but if don't wanna have ESPHome add-on installed, it will work anyways.

> What does not work is the generation of the TFT file as configured in the blueprint.

You don't have to generate the TFT. Just download it from GitHub.

> So I guess my question is: how can I get this nextion service that is used by the blueprint?

There's no Nextion services exposed by the Blueprint. You probably misunderstood this.

What are you trying to achieve?

**Comment by edwardtfn** [comment]
To make it clear, the Blueprint calls the services published by the ESPHome firmware running in the panel.
Based on the calls from the Blueprint, ESPHome will send the proper calls to the Nextion Display, and will get the users activity from the display and send it to the Blueprint either by some of its published entities or by event calls (event named `esphome.nspanel_ha_blueprint`).

**Comment by cmuellner** [comment]
Ah, ok. So the TFT needs to be uploaded only once.
And then the Blueprint just calls ESPhome-services on the NSPanel?

If that's the case, I might have just messed up the TFT upload.

**Comment by cmuellner** [comment]
Found the relevant part in the docs, that seem to confirm this: https://github.com/Blackymas/NSPanel_HA_Blueprint/wiki/(EN)-HowTo---All-important-thing-you-should-know#4-update-tft

> To install a new version of the NSPanel, the Blueprint, ESPHome and the TFT must be updated!

So Blueprint and TFT need to fit together. My misunderstanding was, that I thought the TFT is generated from the blueprint and then uploaded to the device.

Instead the instantiated Blueprint just calls services on the NSPanel (which needs to run the right firmware and has the right TFT). The firmware is expected to change rarely, but the TFT might change with every feature update of the Blueprint so an update of the Blueprint could also require an update of the TFT.

At this point this whole ticket is invalid. Thanks for providing the right pointer!

So my plan is now to upload the TFT again. But it seems I ran into the known issue #1133 as the upload progress drastically decreases after 13%. Uploading the blank TFT worked fine. The download URL is my local HTTP server.

**Comment by edwardtfn** [comment]
> But it seems I ran into the known issue #1133 as the upload progress drastically decreases after 13%. Uploading the blank TFT worked fine. The download URL is my local HTTP server.

Try to use `ref: dev` just to upload the TFT. We are working on a new engine that is available on `dev` only. After uploading, there goes back to `ref: main` otherwise you will have a version mismatch.

**Comment by edwardtfn** [comment]
> The firmware is expected to change rarely, but the TFT might change with every feature update of the Blueprint so an update of the Blueprint could also require an update of the TFT.

Well, the firmware also changes from time to time. Some of the functionalities are fully controlled locally, so it can work even when the Wi-Fi or HA are out, but that requires some of the logic to be in the firmware side.
Usually we release all the 3 parts together. Blueprint + ESPHome (firmware) and TFT.

**Comment by cmuellner** [comment]
Got everything working here.

I attempted to debug the connection reset issue when uploading the TFT file.
The delay until the device was usable in HA was quite annoying, so at one point I switch to the WebIf of the device.
And immediately on the first attempt everything worked as expected (I also disabled the generated automation).
Another idea would be, that I have too many 2.4 GHz networks around me (or other interferences on 2.4 GHz).

Of course, I messed up the Blueprint config as well.
However, after deleting the device and instantiating the automation from the Blueprint again, everything worked.

@edwardtfn: thank you very much for your explanations! I got a quite good understanding of the whole SW stack.
